### PR TITLE
Update generation code for the specific case of geojson

### LINKED ISSUE #393: define type for GeoJSONSourceSpecification data
# Design Proposal: Define type for GeoJSONSourceSpecification data

### General
Developers expect their IDEs to help them write what is needed, often aided by the type system of a package.
If a developer attempts to use a package and gets undesired results, they go looking for solutions in various places (SO, Github, etc), making them less productive.
A well managed type system is critical for developer experience.

### Practical
When attempting to reproduce the following demo,
https://maplibre.org/maplibre-gl-js/docs/examples/multiple-geometries/

...in a modern (Typescript, React) development environment, the sources and layers are not producing an expected visual result. There is no confirmation that `data` or `data.features` are correct because of the use of the `unknown` type for `data`. It becomes one more thing that needs to be verified. Useful typing could avoid that rabbit hole.

```javascript
const sourceId = `source-location-fences-project-${projectId}`
console.log('fences: adding source ' + sourceId)
map.addSource(sourceId, {
  type: 'geojson',
  // HELP NEEDED HERE, DEVELOPERS ARE FLYING BLIND
  data: {
    type: 'FeatureCollection',
    features: locations.map(loc => ({
      type: 'Feature',
      geometry: loc.fence,
    }))
  }
})
const layerId = `layer-location-fences-project-${projectId}`
console.log('fences: adding layer ' + layerId)
map.addLayer({
  id: layerId,
  type: 'fill',
  source: sourceId,
  paint: {
      'fill-color': '#888888',
      'fill-opacity': 0.4
  },
})
```

The type system for `source.data` shows `unknown`, which provides no value for a developer.

## Proposed Change

Improve the type used for `source.data` from `unknown` to `FeatureCollection` from `@types/geojson`.

### Outstanding Questions
1. I do not currently know which [rollup configuration](https://github.com/maplibre/maplibre-style-spec/blob/main/rollup.config.ts) produces the the types.
2. I do not currently know what source file is responsible for the following exported type visible in the `dist/index.d.ts` file.

```
export type GeoJSONSourceSpecification = {
	"type": "geojson";
	"data": unknown;
	"maxzoom"?: number;
	"attribution"?: string;
	"buffer"?: number;
	"filter"?: unknown;
	"tolerance"?: number;
	"cluster"?: boolean;
	"clusterRadius"?: number;
	"clusterMaxZoom"?: number;
	"clusterMinPoints"?: number;
	"clusterProperties"?: unknown;
	"lineMetrics"?: boolean;
	"generateId"?: boolean;
	"promoteId"?: PromoteIdSpecification;
};
```

## API Modifications

No run-time changes. The contract will be better communicated via the improved type.

## Migration Plan and Compatibility

Bump the version number. This feels like a minor version bump to me.

The functionality and feature will be the same, but now users abusing the (previously unstated) contract may see transpile errors.

## Rejected Alternatives

No known alternatives for improving current developer experience.

### PR DESCRIPTION
**Pull request by HarelM**
## Launch Checklist

Fixes #393
- #393

This add an ugly if to the code generation flow to facilitate the fact that the json definitions are not full enough to represent the `data` field in the `GeoJSONSourceSpecification`.
It's a very small fix and I think it adds a lot of value, so I can live with it being dirty I guess...

before:
```ts
export type GeoJSONSourceSpecification = {
	"type": "geojson";
	"data": unknown;
	"maxzoom"?: number;
	"attribution"?: string;
	"buffer"?: number;
	"filter"?: unknown;
	"tolerance"?: number;
	"cluster"?: boolean;
	"clusterRadius"?: number;
	"clusterMaxZoom"?: number;
	"clusterMinPoints"?: number;
	"clusterProperties"?: unknown;
	"lineMetrics"?: boolean;
	"generateId"?: boolean;
	"promoteId"?: PromoteIdSpecification;
};
```
After
```ts
export type GeoJSONSourceSpecification = {
	"type": "geojson";
	"data": GeoJSON.GeoJSON | string;
	"maxzoom"?: number;
	"attribution"?: string;
	"buffer"?: number;
	"filter"?: unknown;
	"tolerance"?: number;
	"cluster"?: boolean;
	"clusterRadius"?: number;
	"clusterMaxZoom"?: number;
	"clusterMinPoints"?: number;
	"clusterProperties"?: unknown;
	"lineMetrics"?: boolean;
	"generateId"?: boolean;
	"promoteId"?: PromoteIdSpecification;
};
```

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,19 +1,30 @@
 ## main
 
+### ✨ Features and improvements
+
 * Add back `migrate`, `validate` and `format` to the exported methods of this package
 * Add terrain to diff method and improve type. This also removes the `operations` from the API [#460](https://github.com/maplibre/maplibre-style-spec/pull/460)
+* Improve the type of `data` in the `GeoJSONSourceSpecification` for typescript.
```
Comment (louwers) [inline CHANGELOG.md:7]:
```suggestion
* Improve the type of `data` in the `GeoJSONSourceSpecification` for TypeScript.
```

**Review by louwers — APPROVED** [review]
This is great!

General points:

- Maybe the file needs to be updated to make clear that it is generating TypeScript.
- I didn't try to run it. Is there any kind of CI in place to validate the output of this script? For example, to ensure that the TypeScript it generates is valid?

**Comment by HarelM** [comment]
Yes, this is a pre-build script to generate typescript files from the v8 json spec file, these files are then used to build the package and are checked during the typescript transpilation.

**Comment by HarelM** [comment]
> Maybe the file needs to be updated to make clear that it is generating TypeScript.

Please propose what you think might help here.
There are two stages that generate typescript, so it might be confusing:
1. This pre-build script
2. The `dts-bundle-generator` which generate the `d.ts` files (the exported types) to be consumed by the user of this package.

**Comment by HarelM** [comment]
Ok, one last PR that I need to handle before releasing a new version which is currently stale and I need to rewrite myself.

**Pull request merged**